File: gps-attendance/android/app/src/main/java/com/gpsattendance/data/ApiService.kt
package com.gpsattendance.data

import retrofit2.http.*

interface ApiService {
    @POST("api/login")
    suspend fun login(@Body request: LoginRequest): LoginResponse

    @POST("api/verify-location")
    suspend fun verifyLocation(
        @Header("Authorization") token: String,
        @Body request: LocationRequest
    ): LocationResponse

    @POST("api/attendance/checkin")
    suspend fun checkIn(
        @Header("Authorization") token: String,
        @Body request: CheckInRequest
    ): CheckInResponse

    @POST("api/attendance/checkout")
    suspend fun checkOut(
        @Header("Authorization") token: String,
        @Body request: CheckOutRequest
    ): CheckOutResponse

    @GET("api/attendance/history")
    suspend fun getHistory(
        @Header("Authorization") token: String,
        @Query("emp_id") empId: Int
    ): HistoryResponse
}

data class LoginRequest(val username: String, val password: String)

data class LoginResponse(
    val token: String,
    val user: UserInfo
)

data class UserInfo(
    val id: Int,
    val username: String,
    val role: String,
    val employee_id: Int,
    val employee_name: String
)

data class LocationRequest(
    val employee_id: Int,
    val latitude: Double,
    val longitude: Double,
    val site_id: Int
)

data class LocationResponse(
    val valid: Boolean,
    val distance_meters: Double,
    val allowed_radius: Int,
    val site_name: String
)

data class CheckInRequest(
    val employee_id: Int,
    val site_id: Int,
    val latitude: Double,
    val longitude: Double
)

data class CheckInResponse(
    val success: Boolean,
    val attendance_id: Int,
    val check_in_time: String,
    val location_valid: Boolean,
    val distance_meters: Double
)

data class CheckOutRequest(
    val employee_id: Int,
    val attendance_id: Int,
    val latitude: Double,
    val longitude: Double
)

data class CheckOutResponse(
    val success: Boolean,
    val check_out_time: String,
    val location_valid: Boolean,
    val distance_meters: Double
)

data class HistoryResponse(
    val records: List<AttendanceHistoryRecord>
)

data class AttendanceHistoryRecord(
    val id: Int,
    val employee_id: Int,
    val site_id: Int,
    val site_name: String,
    val check_in_time: String,
    val check_out_time: String?,
    val check_in_valid: Boolean,
    val check_out_valid: Boolean?
)

File: gps-attendance/android/lib/services/api_service.dart
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

class ApiService {
  static const String baseUrl = 'http://YOUR_API_HOST/gps-attendance/api';

  static Future<Map<String, String>> _headers() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('auth_token') ?? '';
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  static Future<Map<String, dynamic>> login(String username, String password) async {
    final response = await http.post(
      Uri.parse('$baseUrl/login'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'username': username, 'password': password}),
    );
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> checkin(double lat, double lng, String siteId) async {
    final headers = await _headers();
    final response = await http.post(
      Uri.parse('$baseUrl/attendance/checkin'),
      headers: headers,
      body: jsonEncode({
        'latitude': lat,
        'longitude': lng,
        'site_id': siteId,
        'check_in_time': DateTime.now().toIso8601String(),
      }),
    );
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> checkout(double lat, double lng) async {
    final headers = await _headers();
    final response = await http.post(
      Uri.parse('$baseUrl/attendance/checkout'),
      headers: headers,
      body: jsonEncode({
        'latitude': lat,
        'longitude': lng,
        'check_out_time': DateTime.now().toIso8601String(),
      }),
    );
    return jsonDecode(response.body);
  }

  static Future<List<dynamic>> getHistory(String empId, {String? from, String? to}) async {
    final headers = await _headers();
    String url = '$baseUrl/attendance/history?emp_id=$empId';
    if (from != null) url += '&from=$from';
    if (to != null) url += '&to=$to';
    final response = await http.get(Uri.parse(url), headers: headers);
    return jsonDecode(response.body);
  }

  static Future<Map<String, dynamic>> verifyLocation(double lat, double lng, String siteId) async {
    final headers = await _headers();
    final response = await http.post(
      Uri.parse('$baseUrl/verify-location'),
      headers: headers,
      body: jsonEncode({
        'latitude': lat,
        'longitude': lng,
        'site_id': siteId,
      }),
    );
    return jsonDecode(response.body);
  }
}